packages/block-editor/src/components/provider/test/use-block-sync.js

```diff
@@ -362,4 +362,77 @@ describe( 'useBlockSync hook', () => {
 		);
 		expect( onInput ).not.toHaveBeenCalled();
 	} );
+
+	it( 'should use fresh callbacks if onChange/onInput have been updated', async () => {
+		const fakeBlocks = [
+			{ clientId: 'a', innerBlocks: [], attributes: { foo: 1 } },
+		];
+		const onChange1 = jest.fn();
+		const onInput = jest.fn();
+
+		let registry;
+		const setRegistry = ( reg ) => {
+			registry = reg;
+		};
+		let root;
+		await act( async () => {
+			root = create(
+				<TestWrapper
+					setRegistry={ setRegistry }
+					value={ fakeBlocks }
+					onChange={ onChange1 }
+					onInput={ onInput }
+				/>
+			);
+		} );
+
+		// Create a persistent change.
+		registry
+			.dispatch( 'core/block-editor' )
+			.updateBlockAttributes( 'a', { foo: 2 } );
+
+		const updatedBlocks1 = [
+			{ clientId: 'a', innerBlocks: [], attributes: { foo: 2 } },
+		];
+
+		expect( onChange1 ).toHaveBeenCalledWith( updatedBlocks1, {
+			selectionEnd: {},
+			selectionStart: {},
+		} );
```

Comment:
Does this update and assertion make sense for this test?  I think the bug and change aren't specific to the first onChange having to be called first.  On one hand it adds an extra assertion and some safety in case there are other issues with the initial onChange... on the other it biases the the following assertions based on the original onChange having been called (although in this case I wouldn't expect that to cause any problems).  

Philosophically speaking -  maybe this should be a separate test, or just removed if it is already covered elsewhere?  🤔

Reply:
> maybe this should be a separate test

Thankfully, there is a test that covers this by itself :)

I wanted to add this to test the whole lifecycle of a callback becoming stale -- basically, you have one callback and works, and then you switch to another callback which also works (and then makes the first callback stop working).

I could see a case for just removing this assertion, but I feel like the test is more robust without it. But I think it also makes sense to add a test for the scenario when there has not been a change first